File: tools/semantic_ir/statements.py
"""STUNIR Semantic IR Statement Nodes.

DO-178C Level A Compliant
Statement node definitions.
"""

from typing import List, Optional, Union
from pydantic import BaseModel, Field, conint

from .ir_types import (
    NodeID, IRName, IRNodeKind,
    StorageClass, MutabilityKind
)
from .nodes import IRNodeBase, TypeReference


class StatementNode(IRNodeBase):
    """Base statement node."""
    pass


class BlockStmt(StatementNode):
    """Block statement."""
    kind: IRNodeKind = IRNodeKind.BLOCK_STMT
    statements: List[Union[NodeID, "StatementNode"]] = Field(default_factory=list)
    scope_id: Optional[IRName] = None


class ExprStmt(StatementNode):
    """Expression statement."""
    kind: IRNodeKind = IRNodeKind.EXPR_STMT
    expression: Union[NodeID, "ExpressionNode"]


class IfStmt(StatementNode):
    """If statement."""
    kind: IRNodeKind = IRNodeKind.IF_STMT
    condition: Union[NodeID, "ExpressionNode"]
    then_branch: Union[NodeID, "StatementNode"]
    else_branch: Optional[Union[NodeID, "StatementNode"]] = None


class WhileStmt(StatementNode):
    """While loop statement."""
    kind: IRNodeKind = IRNodeKind.WHILE_STMT
    condition: Union[NodeID, "ExpressionNode"]
    body: Union[NodeID, "StatementNode"]
    loop_bound: conint(ge=0) = 0  # 0 means unbounded
    unroll: bool = False


class ForStmt(StatementNode):
    """For loop statement."""
    kind: IRNodeKind = IRNodeKind.FOR_STMT
    init: Union[NodeID, "StatementNode"]
    condition: Union[NodeID, "ExpressionNode"]
    increment: Union[NodeID, "StatementNode"]
    body: Union[NodeID, "StatementNode"]
    loop_bound: conint(ge=0) = 0
    unroll: bool = False
    vectorize: bool = False


class ReturnStmt(StatementNode):
    """Return statement."""
    kind: IRNodeKind = IRNodeKind.RETURN_STMT
    value: Optional[Union[NodeID, "ExpressionNode"]] = None


class BreakStmt(StatementNode):
    """Break statement."""
    kind: IRNodeKind = IRNodeKind.BREAK_STMT


class ContinueStmt(StatementNode):
    """Continue statement."""
    kind: IRNodeKind = IRNodeKind.CONTINUE_STMT


class VarDeclStmt(StatementNode):
    """Variable declaration statement."""
    kind: IRNodeKind = IRNodeKind.VAR_DECL_STMT
    name: IRName
    var_type: Optional[TypeReference] = None
    initializer: Optional[Union[NodeID, "ExpressionNode"]] = None
    storage: StorageClass = StorageClass.AUTO
    mutability: MutabilityKind = MutabilityKind.MUTABLE


class AssignStmt(StatementNode):
    """Assignment statement."""
    kind: IRNodeKind = IRNodeKind.ASSIGN_STMT
    target: Union[NodeID, "ExpressionNode"]
    value: Union[NodeID, "ExpressionNode"]


# Enable forward references
from .expressions import ExpressionNode  # noqa: E402

BlockStmt.model_rebuild()
ExprStmt.model_rebuild()
IfStmt.model_rebuild()
WhileStmt.model_rebuild()
ForStmt.model_rebuild()
ReturnStmt.model_rebuild()
VarDeclStmt.model_rebuild()
AssignStmt.model_rebuild()

File: tools/rust/semantic_ir/src/statements.rs
//! STUNIR Semantic IR Statement Nodes
//!
//! Statement node definitions.

use serde::{Deserialize, Serialize};

use crate::types::*;
use crate::nodes::*;

/// Statement node
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct StatementNode {
    #[serde(flatten)]
    pub base: IRNodeBase,
}

/// Block statement
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct BlockStmt {
    #[serde(flatten)]
    pub base: StatementNode,
    pub statements: Vec<NodeID>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub scope_id: Option<IRName>,
}

/// Expression statement
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ExprStmt {
    #[serde(flatten)]
    pub base: StatementNode,
    pub expression: NodeID,
}

/// If statement
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct IfStmt {
    #[serde(flatten)]
    pub base: StatementNode,
    pub condition: NodeID,
    pub then_branch: NodeID,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub else_branch: Option<NodeID>,
}

/// While statement
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct WhileStmt {
    #[serde(flatten)]
    pub base: StatementNode,
    pub condition: NodeID,
    pub body: NodeID,
    #[serde(default)]
    pub loop_bound: u32,
    #[serde(default)]
    pub unroll: bool,
}

/// For statement
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ForStmt {
    #[serde(flatten)]
    pub base: StatementNode,
    pub init: NodeID,
    pub condition: NodeID,
    pub increment: NodeID,
    pub body: NodeID,
    #[serde(default)]
    pub loop_bound: u32,
    #[serde(default)]
    pub unroll: bool,
    #[serde(default)]
    pub vectorize: bool,
}

/// Return statement
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ReturnStmt {
    #[serde(flatten)]
    pub base: StatementNode,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub value: Option<NodeID>,
}

/// Break statement
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct BreakStmt {
    #[serde(flatten)]
    pub base: StatementNode,
}

/// Continue statement
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ContinueStmt {
    #[serde(flatten)]
    pub base: StatementNode,
}

/// Variable declaration statement
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct VarDeclStmt {
    #[serde(flatten)]
    pub base: StatementNode,
    pub name: IRName,
    #[serde(skip_serializing_if = "Option::is_none", rename = "var_type")]
    pub var_type: Option<TypeReference>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub initializer: Option<NodeID>,
    #[serde(default)]
    pub storage: StorageClass,
    #[serde(default)]
    pub mutability: MutabilityKind,
}

impl Default for StorageClass {
    fn default() -> Self { StorageClass::Auto }
}

impl Default for MutabilityKind {
    fn default() -> Self { MutabilityKind::Mutable }
}

/// Assignment statement
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct AssignStmt {
    #[serde(flatten)]
    pub base: StatementNode,
    pub target: NodeID,
    pub value: NodeID,
}
